Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+///
+/// \file    DatDecoder.cxx
+/// \author  Andrea Ferrero
+///
+/// \brief Implementation of a data processor to run the raw decoding
+///
+
+#include <random>
+#include <iostream>
+#include <fstream>
+#include <stdexcept>
+#include "Framework/CallbackService.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/ControlService.h"
+#include "Framework/DataProcessorSpec.h"
+#include "Framework/Lifetime.h"
+#include "Framework/Output.h"
+#include "Framework/Task.h"
+#include "Framework/WorkflowSpec.h"
+
+#include "DPLUtils/DPLRawParser.h"
+#include "MCHBase/Digit.h"
+#include "Headers/RAWDataHeader.h"
+#include "DetectorsRaw/RDHUtils.h"
+#include "MCHRawCommon/DataFormats.h"
+#include "MCHRawDecoder/PageDecoder.h"
+#include "MCHRawElecMap/Mapper.h"
+#include "MCHMappingInterface/Segmentation.h"
+#include "MCHWorkflow/DataDecoder.h"
+
+static bool mPrint = false;
+
+namespace o2
+{
+namespace mch
+{
+namespace raw
+{
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::mch::mapping;
+using RDHv4 = o2::header::RAWDataHeaderV4;
+
+static std::array<int, 64> refManu2ds_st345_v5 = {
+  63, 62, 61, 60, 59, 57, 56, 53, 51, 50, 47, 45, 44, 41, 38, 35,
+  36, 33, 34, 37, 32, 39, 40, 42, 43, 46, 48, 49, 52, 54, 55, 58,
+  7, 8, 5, 2, 6, 1, 3, 0, 4, 9, 10, 15, 17, 18, 22, 25,
+  31, 30, 29, 28, 27, 26, 24, 23, 20, 21, 16, 19, 12, 14, 11, 13};
+
+static std::array<int, 64> refManu2ds_st345_v2 = {
+  62, 61, 63, 60, 59, 55, 58, 57, 56, 54, 50, 46, 42, 39, 37, 41,
+  35, 36, 33, 34, 32, 38, 43, 40, 45, 44, 47, 48, 49, 52, 51, 53,
+  7, 6, 5, 4, 2, 3, 1, 0, 9, 11, 13, 15, 17, 19, 21, 23,
+  31, 30, 29, 28, 27, 26, 25, 24, 22, 20, 18, 16, 14, 12, 10, 8};
+
+#define refManu2ds_st345 refManu2ds_st345_v5
+
+static std::array<int, 64> refDs2manu_st345;
+
+static int manu2ds(int i)
+{
+  return refManu2ds_st345[i];
+}
+
+static int ds2manu(int i)
+{
+  return refDs2manu_st345[i];
+}
+
+class MergerDigit
+{
+ public:
+  MergerDigit() = default;
+  ~MergerDigit() = default;
+
+  o2::mch::Digit digit;
+  o2::mch::Digit::Time stopTime;
+  bool merged = {false};
+  int solarId = {-1};
+  int dsAddr = {-1};
+  int chAddr = {-1};
+};
+
+//_________________________________________________________________
+struct MergerBuffer {
+  std::vector<MergerDigit> digits;
+  uint32_t orbit = {0};
+};
+
+//_________________________________________________________________
+class FeeIdMerger
+{
+  MergerBuffer buffers[2];
```

Comment:
Instead of having to juggle with two buffers indices, why not define two actual buffers named currentBuffer and previousBuffer and use a swap method to revert them when needed (if I understand correctly is what's done below in L230 ?). Then you can simplify and remove the BufId (and their get methods) altogether

Reply:
Good suggestion. I will change the code to use `std::swap`.